### PR TITLE
Test real token flavors through upgradeability with generic eternal storage

### PR DESCRIPTION
**Pull request by facuspagnuolo**
The idea was to demonstrate that we can go through different token behaviors using upgradeability, in this case with generic eternal storage, taking some of the token flavors most chosen by the industry:
- V0: Standard token
- V1: Mintable (ownable) token
- V2: Burnable token
- V3: Pausable token

### CODE REVIEW

File: upgradeability_using_generic_eternal_storage/contracts/test/OwnedToken.sol [hunk 1]
```diff
@@ -34,7 +34,7 @@ contract OwnedToken is EternalStorage {
    * @dev Allows the current owner to transfer control of the contract to a newOwner.
    * @param newOwner The address to transfer ownership to.
    */
-  function transferOwnership(address newOwner) public onlyTokenOwner {
+  function transferTokenOwnership(address newOwner) public onlyTokenOwner {
```
Comment (maraoz) [inline upgradeability_using_generic_eternal_storage/contracts/test/OwnedToken.sol]: why can't you use vanilla `Ownable` for this?

File: upgradeability_using_generic_eternal_storage/contracts/test/Token_V2.sol [hunk 2]
```diff
@@ -0,0 +1,21 @@
+pragma solidity ^0.4.18;
+
+import './Token_V1.sol';
+
+/**
+ * @title Token_V2
+ * @dev Version 2 of a token to show upgradeability.
+ * The idea here is to extend a token behaviour providing burnable functionalities as opposed to version 1
```
Comment (maraoz) [inline upgradeability_using_generic_eternal_storage/contracts/test/Token_V2.sol]: "as opposed to" => "in addition to what's provided in"

File: upgradeability_using_generic_eternal_storage/contracts/test/Token_V3.sol [hunk 3]
```diff
@@ -0,0 +1,87 @@
+pragma solidity ^0.4.18;
+
+import './Token_V2.sol';
+
+/**
+ * @title Token_V2
+ * @dev Version 2 of a token to show upgradeability.
+ * The idea here is to extend a token behaviour providing pausable functionalities as opposed to version 2
```
Comment (maraoz) [inline upgradeability_using_generic_eternal_storage/contracts/test/Token_V3.sol]: "as opposed to" => "in addition to what's provided in"

File: upgradeability_using_generic_eternal_storage/contracts/test/OwnedToken.sol (hunk 1 above)
Comment (facuspagnuolo) [reply upgradeability_using_generic_eternal_storage/contracts/test/OwnedToken.sol]: The problem is that we need to handle two different ownerships: one for upgradeability purpose and another one for the token. This means, having two different ways of transferring ownerships. We could have used `transferOwnership` for one of them, but it wouldn't have been clear enough to use a custom function name in one of the contracts and a generic one in the other.

Bear in mind that it is not possible to use `trasnferOwnership` for both contracts given we are using the fallback function of the `Proxy` to delegate requests to the last upgraded token behavior. 

On the other hand, we could have put the `OwnedToken` behavior inside the `Token_V1` itself. But I tried to followed the inheritance approach as if we were using OZ. Although, you will see that [`Token_V3`](https://github.com/zeppelinos/labs/pull/9/files#diff-95e31b5b2498433fe07c9ac815b8bfefR11) is a pausable token that implements `Pausable` behavior itself, I didn't model a `Pausable` contract and inherited from it. We probably may want to choose between one of the two approaches, but I wanted to give an example to show both ways and discuss about them.

**Comment by facuspagnuolo** [comment]
As we discussed offline, we moved the `OwnedToken` to a vanilla `Ownable` contract given that the token should not be aware of a `Proxy`, it should work as if we were using OZ.

**Pull request merged**